### ISSUE TITLE
[dynamo] OpenVoice VITS: Tensor.item() in @torch.jit.script helper (2 models; reversed(ModuleList) gap from sweep no longer reproduces on nightly)

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#14]
**Repro status:** Reproduces on torch 2.12.0.dev20260407+cu128 (nightly, verified 2026-05-09 ~17:58 UTC) — for the **`Tensor.item()` gap**. The original sweep evidence ALSO listed `reversed(ModuleList)` as a separate gap; that gap appears to be **fixed in current nightly** (probe on real ResidualCouplingBlock returns 0 graph breaks; trivial 2-Linear ModuleList with `for x in reversed(ModuleList):` compiles cleanly with `fullgraph=True`).

> **Multi-gap issue note:** This issue tracks several Dynamo gaps in the OpenVoice VITS pipeline. The MRE below covers the still-firing `Tensor.item()` gap inside a `@torch.jit.script` helper. The originally-prominent `reversed(ModuleList)` gap is now fixed upstream; aten.nonzero / data-dep branching gaps (also listed) likely have similar status — recommend the maintainer evaluate which gaps still need attention before scoping a fix.

## Original failure report

2 OpenVoice VITS components produce graph breaks. The dominant remaining gap is `Tensor.item()` at `corpora/custom-models/_sources/openvoice_synthesizertrn/openvoice/commons.py:104` — inside a `@torch.jit.script` helper that does tensor-indexed slicing (`x[:, :n, :]` where `n = n_channels[0]`).

**Test command (sweep):** `python tools/run_experiment.py sweep ...` (cohort: OpenVoice-SynthesizerTrn, OpenVoice-SynthesizerTrn-infer, nightly 2026-04-26).

**Symptom captured:** `Unsupported Tensor.item() call with capture_scalar_outputs=False`. Cited stack frame: `commons.py:104`.

### Affected Models

| Model | Modes | Breaks (sweep) |
|---|---|---|
| OpenVoice-SynthesizerTrn | eval | 8-12 |
| OpenVoice-SynthesizerTrn-infer | eval | 23-26 |

<details><summary>Verification signal</summary>

`{"kind": "stderr_contains", "fragment": "Unsupported Tensor.item() call with capture_scalar_outputs=False"}`
</details>

## Pattern

Cited call site at `corpora/custom-models/_sources/openvoice_synthesizertrn/openvoice/commons.py:104`:

```text
@torch.jit.script
def fused_add_tanh_sigmoid_multiply(input_a, input_b, n_channels):
    n_channels_int = n_channels[0]  # tensor-indexed extraction
    in_act = input_a + input_b
    t_act = torch.tanh(in_act[:, :n_channels_int, :])  # uses extracted scalar
    ...
```

Inside `@torch.jit.script`, the tensor-indexed extraction `n_channels[0]` returns a 0-dim tensor; subsequent slicing requires Dynamo to materialize it via `.item()` — which fails under `capture_scalar_outputs=False`.

## Minimal reproducer (MRE)

24 lines, **no openvoice dependency** — uses only `torch.jit.script` + a tiny indexed-slicing helper.

```python repro=true
"""MRE: graph break from `Tensor.item()` triggered by tensor-indexed slicing
inside a `@torch.jit.script` helper.
"""
import os
os.environ.setdefault("TORCH_LOGS", "graph_breaks")

import torch
import torch.nn as nn


@torch.jit.script
def slice_with_tensor_index(x, n_channels):
    n = n_channels[0]
    return torch.tanh(x[:, :n, :])


class M(nn.Module):
    def __init__(self):
        super().__init__()
        self.hidden_channels = 4

    def forward(self, x):
        n_channels_tensor = torch.IntTensor([self.hidden_channels])
        return slice_with_tensor_index(x, n_channels_tensor)


m = M()
x = torch.randn(1, 8, 4)
compiled = torch.compile(m, backend="eager")
compiled(x)
```

<details><summary>Verification signal (MRE)</summary>

`{"kind": "stderr_contains", "fragment": "Unsupported Tensor.item() call with capture_scalar_outputs=False"}`
</details>

## Source

- Sweep results: `sweep_results/nightly/2026-04-26/explain_results.json` (OpenVoice rows)
- **Provenance anchor:** `corpora/custom-models/_sources/openvoice_synthesizertrn/openvoice/commons.py:104`
- Built off PR pytorch/pytorch#181552
- Transformers: 5.6.2

---
<sub>Body rewritten 2026-05-09 by file-issue + mre subagent dogfood. **Important:** the original `reversed(ModuleList)` gap from the 2026-04-26 sweep no longer reproduces on current nightly — appears to have been fixed upstream. MRE covers the still-firing `Tensor.item()` inside `@torch.jit.script` helper.</sub>

**Comment by penguinwu** [comment]
### Nightly Validation (torch 2.12.0.dev20260407 vs 2.11.0 stable)

**Status: Persists on nightly.** Custom model worker outputs show graph break logs on nightly (worker parsing issue, not a fix). Root causes: `Tensor.item()`, `aten.nonzero`, data-dependent branching — all fundamental Dynamo limitations.

| Model | v2.11 | Nightly | Notes |
|-------|-------|---------|-------|
| OpenVoice-SynthesizerTrn | 7 breaks (eval) | graph_break (worker parse error) | Tensor.item() |
| OpenVoice-SynthesizerTrn-infer | 22 breaks (eval) | graph_break (worker parse error) | nonzero + data-dep |

**Conclusion: Issue persists on nightly.**

**Comment by penguinwu** [comment]
**Note on `Tensor.item()` breaks:** With `fullgraph=True`, `capture_scalar_outputs` is effectively True, so item() calls are captured into the graph (not graph-broken). The item()-related breaks listed here are from the explain pass which runs without fullgraph. Users compiling with `fullgraph=True` may see different (potentially fewer) breaks.

We've updated our explain pass to set `capture_scalar_outputs=True` and `capture_dynamic_output_shape_ops=True` for consistency with fullgraph behavior. Future explain data will no longer report item() as a separate graph break.

Keeping this issue open — non-item breaks (aten.nonzero, data-dependent branching, rand_slice_segments) still apply.
